[TIMOB-9366] Android: Added "radial" gradient support

jquick-axway opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-9366

**Notes:**
- This feature is intended as an alternative for devs who are drawing circles/rings via a view's `borderRadius` property, which has known issues on Google's end.
- Supports all gradient properties except `endPoint`, which on iOS is used to smear the circular gradient from the start to end point. Android does not support this.
- Updated API docs to indicate Android's support and to better clarify the offset range for linear and radial gradients.
- The unit test written for PR #9819 covers gradient testing for both Android and iOS.

**Test:**
1. Build and run "GradientRadialTest.js" attached to [TIMOB-9366](https://jira.appcelerator.org/browse/TIMOB-9366) on Android.
2. On startup, verify that you see a circle where the outer edge is blue and the center is red.
3. Drag the "Start Radius" slider to the middle.
4. Verify that the circle becomes a ring. _(ie: The center is transparent)_
5. Tap the "Back Fill End" switch to enable it.
6. Verify that the area outside of the circle is now blue.
7. Tap the "Back Fill Start" switch to enable it.
8. Verify that the center of the circle is now solid red. _(ie: No longer transparent.)_
9. Tap the "Back Fill End" switch to disable it.
10. Drag the "End Radius" slider to the left, lower than the start radius position. _(ie: You want the end radius to be less than the start radius.)_
11. Verify that the inner part of the circle is now blue with a transparent center and the part outside of the circle is solid red.
12. Touch-drag the circle with your finger and verify that the center of the circle moves with your finger.

jquick-axway committed:
[TIMOB-9366] Android: Added "radial" gradient support to view's "backgroundGradient" property

jquick-axway committed:
Merge branch 'master' into TIMOB-9366

build commented:
<!--
  2 failure:  :disappointed_rel..., :microscope: Ther...
  0 warning: 
  1 messages
  
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/jquick-axway/titanium_mobile/blob/TIMOB-9366/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.1.0 test /Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9815-AELG27D6DZ7GEIZIMNNJTA6IT36ZVOLIXZ26F4QLGBG7ZVPRDCBA
> grunt

Running "appcJs:src" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9815-AELG27D6DZ7GEIZIMNNJTA6IT36ZVOLIXZ26F4QLGBG7ZVPRDCBA/android/cli/commands/_buildModule.js
  554:18  warning  Found non-literal argument to RegExp Constructor  security/detect-non-literal-regexp

✖ 1 problem (0 errors, 1 warning)


Running "continue:on" task

Running "retire:js" (retire) task
Ignoring {"paths":["growl"],"descriptors":[]}
Downloading https://raw.github.com/RetireJS/retire.js/master/repository/jsrepository.json ...
(node:81277) [DEP0013] DeprecationWarning: Calling an asynchronous function without callback is deprecated.
Checking: dangerfile.js
Checking: Gruntfile.js
Checking: apidoc/docgen.js
Checking: apidoc/lib/addon_generator.js
Checking: apidoc/lib/changes_generator.js
Checking: apidoc/lib/common.js
Checking: apidoc/lib/html_generator.js
Checking: apidoc/lib/jsca_generator.js
Checking: apidoc/lib/jsduck_generator.js
Checking: apidoc/lib/json_generator.js
Checking: apidoc/lib/modulehtml_generator.js
Checking: apidoc/lib/parity_generator.js
Checking: apidoc/lib/solr_generator.js
Checking: apidoc/validate.js
Checking: build/android.js
Checking: build/androidsdk.js
Checking: build/ant.js
Checking: build/docs.js
Checking: build/git.js
Checking: build/ios.js
Checking: build/packager.js
Checking: build/scons-build.js
Checking: build/scons-clean.js
Checking: build/scons-cleanbuild.js
Checking: build/scons-install.js
Checking: build/scons-modules-integrity.js
Checking: build/scons-package.js
Checking: build/scons-ssri.js
Checking: build/scons-test.js
Checking: build/scons-update-node-deps.js
Checking: build/scons.js
Checking: build/titanium-sdk/lib/titanium.js
Checking: build/utils.js
Checking: build/windows.js
Checking: cli/commands/build.js
Checking: cli/commands/clean.js
Checking: cli/commands/create.js
Checking: cli/commands/project.js
Checking: cli/hooks/hyperloop-fix.js
Checking: cli/lib/creator.js
Checking: cli/lib/creators/app.js
Checking: cli/lib/creators/applewatch.js
Checking: cli/lib/creators/module.js
Checking: cli/lib/info.js
Checking: android/cli/commands/_build.js
Checking: android/cli/commands/_buildModule.js
Checking: android/cli/hooks/aar-transform.js
Checking: android/cli/hooks/package.js
Checking: android/cli/hooks/run.js
Checking: android/cli/lib/AndroidManifest.js
Checking: android/cli/lib/base-builder.js
Checking: android/cli/lib/detect.js
Checking: android/cli/lib/info.js
Checking: android/cli/tests/run.js
Checking: android/cli/tests/test-androidmanifest.js
Checking: iphone/cli/commands/_build.js
Checking: iphone/cli/commands/_buildModule.js
Checking: iphone/cli/hooks/frameworks.js
Checking: iphone/cli/hooks/hyperloop.js
Checking: iphone/cli/hooks/install.js
Checking: iphone/cli/hooks/package.js
Checking: iphone/cli/hooks/run.js
Checking: iphone/cli/lib/info.js
Checking: tests/Resources/ti.example.addontest.js
Checking: tests/Resources/ti.utils.test.js
No vulnerabilities found.

Running "retire:node" (retire) task
Ignoring {"paths":["growl"],"descriptors":[]}
Downloading https://raw.github.com/RetireJS/retire.js/master/repository/npmrepository.json ...
(node:81277) [DEP0013] DeprecationWarning: Calling an asynchronous function without callback is deprecated.
Checking: .
Missing version for babel-eslint. Need to run npm install ?
Missing version for eslint-plugin-jsx-a11y. Need to run npm install ?
Missing version for eslint-plugin-react. Need to run npm install ?
Missing version for babel-eslint. Need to run npm install ?
Missing version for eslint-plugin-jsx-a11y. Need to run npm install ?
Missing version for eslint-plugin-react. Need to run npm install ?
No vulnerabilities found.

Running "continue:off" task

Running "ios_format:src" (ios_format) task

Running "android_format:src" (android_format) task
Fatal error: Error: Formatting incorrect on "android/titanium/src/java/org/appcelerator/titanium/view/TiGradientDrawable.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='7785' length='0'> </replacement>
<replacement offset='7945' length='0'> </replacement>
<replacement offset='9047' length='0'> </replacement>
<replacement offset='9063' length='0'> </replacement>
<replacement offset='9693' length='0'> </replacement>
<replacement offset='13680' length='0'> </replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

hansemannn committed:
Update TiGradientDrawable.java

hansemannn commented:
Hey @jquick-axway, I've made some unit-tests:
```js
'use strict';
var should = require('./utilities/assertions'),
	utilities = require('./utilities/utilities');

describe('Titanium.UI.View', function () {
	var win,
		didFocus = false,
		didPostLayout = false;

	this.timeout(5000);

	beforeEach(function () {
		didFocus = false;
		didPostLayout = false;
	});

	afterEach(function () {
		if (win) {
			win.close();
		}
		win = null;
	});

	it('backgroundGradient (radial)', function (finish) {
		this.timeout(10000);

		win = Titanium.UI.createWindow({
		    backgroundColor: '#fff'
		});

		var gradient = {
				type: 'radial',
				startPoint: { x: 50, y: 50 },
				colors: [ 'red', 'blue'],
				startRadius: '90%',
				endRadius: 0
		};

		// iOS/Windows-only properties
		if (!utilities.isAndroid()) {
			gradient.endPoint = { x: 50, y: 50 };
			gradient.backfillStart = true;
		}

		var radialGradient = Ti.UI.createView({
		    width: 100,
		    height: 100,
		    backgroundGradient: gradient
		});

		win.addEventListener('open', function() {
			// general API
			should(radialGradient.backgroundGradient).be.an.Object;

			// type
			should(radialGradient.backgroundGradient.type).be.a.String;
			should(radialGradient.backgroundGradient.type).eql('radial');

			// startPoint
			should(radialGradient.backgroundGradient.startPoint).be.an.Object;
			should(radialGradient.backgroundGradient.startPoint.x).be.a.Number;
			should(radialGradient.backgroundGradient.startPoint.y).be.a.Number;
			should(radialGradient.backgroundGradient.startPoint.x).eql(50);
			should(radialGradient.backgroundGradient.startPoint.y).eql(50);

			// endPoint
			should(radialGradient.backgroundGradient.endPoint).be.an.Object;
			should(radialGradient.backgroundGradient.endPoint.x).be.a.Number;
			should(radialGradient.backgroundGradient.endPoint.y).be.a.Number;
			should(radialGradient.backgroundGradient.endPoint.x).eql(50);
			should(radialGradient.backgroundGradient.endPoint.y).eql(50);

			// backfillStart
			should(radialGradient.backgroundGradient.backfillStart).be.a.Boolean;
			should(radialGradient.backgroundGradient.backfillStart).eql(true);

			// colors
			should(radialGradient.backgroundGradient.colors).be.an.Array;
			should(radialGradient.backgroundGradient.colors[0]).eql('red');
			should(radialGradient.backgroundGradient.colors[1]).eql('blue');

			// startRadius
			should(radialGradient.backgroundGradient.startRadius).eql('90%');

			// endRadius
			should(radialGradient.backgroundGradient.endRadius).eql(0);
		});

		win.add(radialGradient);		
		win.open();
	});

	it('backgroundGradient (linear)', function (finish) {
		this.timeout(10000);

		win = Titanium.UI.createWindow({
		    backgroundColor: '#fff'
		});

		var linearGradient = Ti.UI.createView({
		    width: 100,
		    height: 100,
		    backgroundGradient: {
					type: 'linear',
					startPoint: { x: '0%', y: '50%' },
					endPoint: { x: '100%', y: '100%' },
					colors: [ { color: 'red', offset: 0.0 }, { color: 'blue', offset: 0.25 }, { color: 'red', offset: 1.0 } ],
				}
		});

		win.addEventListener('open', function() {
			// general API
			should(linearGradient.backgroundGradient).be.an.Object;

			// type
			should(linearGradient.backgroundGradient.type).be.a.String;
			should(linearGradient.backgroundGradient.type).eql('linear');

			// startPoint
			should(linearGradient.backgroundGradient.startPoint).be.an.Object;
			should(linearGradient.backgroundGradient.startPoint.x).be.a.String;
			should(linearGradient.backgroundGradient.startPoint.y).be.a.String;
			should(linearGradient.backgroundGradient.startPoint.x).eql('0%');
			should(linearGradient.backgroundGradient.startPoint.y).eql('50%');

			// endPoint
			should(linearGradient.backgroundGradient.endPoint).be.an.Object;
			should(linearGradient.backgroundGradient.endPoint.x).be.a.String;
			should(linearGradient.backgroundGradient.endPoint.y).be.a.String;
			should(linearGradient.backgroundGradient.endPoint.x).eql('100%');
			should(linearGradient.backgroundGradient.endPoint.y).eql('100%');

			// colors
			should(linearGradient.backgroundGradient.colors).be.an.Array;
			linearGradient.backgroundGradient.colors.forEach(function (colorObject) {
				should(colorObject).be.an.Object;
				should(colorObject.color).be.a.String;
				should(colorObject.offset).be.a.Number;
			});
		});
		
		win.add(linearGradient);		
		win.open();
	});
});
```
Note: The `should(linearGradient.backgroundGradient.colors).be.an.Array;` test was broken before due to missing proxy getters on iOS. I fixed them as part of an own commit that may be merged in this branch. Also, @infosia, do we support all this on Windows already or should I skip them?

hansemannn committed:
Merge branch 'master' into TIMOB-9366

hansemannn commented:
Added the unit-tests to #9819. If we merge that one before, we don't need to add it here as well.

jquick-axway commented:
Thanks @hansemannn. And I'm not being lazy here. Just trying to catch up with some high priority Android issues first. I'll write up a test for this PR later this week.
_(I've already done extensive testing on my end.)_

hansemannn commented:
Totally fine, just wanted to help :-).

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/view/TiGradientDrawable.java:
```diff
@@ -74,21 +77,187 @@ public TiGradientDrawable(View view, KrollDict properties)
-		if (startRadius == null) {
-			startRadius = DEFAULT_RADIUS;
+		// Fetch the start/end radius values for a "radial" gradient.
+		this.startRadius = TiConvert.toTiDimension(properties, "startRadius", TiDimension.TYPE_UNDEFINED);
```
Minor, but here's a good opportunity to define `"startRadius"` and `"endRadius"` as constants

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/view/TiGradientDrawable.java:
```diff
@@ -74,21 +77,187 @@ public TiGradientDrawable(View view, KrollDict properties)
 		}
 
+		// Fetch the fill properties for a "radial" gradient.
+		this.isBackFillingStart = TiConvert.toBoolean(properties, "backfillStart", false);
```
Same with these properties

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/view/TiGradientDrawable.java:
```diff
@@ -98,33 +267,50 @@ public GradientType getGradientType()
+					if (this.offsets == null) {
+						this.offsets = new float[colors.length];
+					}
+					float offset = TiConvert.toFloat(colorRefObject, "offset", -1);
```
Use `OFFSET_KEY` instead of `"offset"`

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/view/TiGradientDrawable.java:
```diff
@@ -27,35 +29,36 @@
+
 		// Determine which type of gradient is being used.
 		// Supported types are 'linear' and 'radial'.
 		String type = properties.optString("type", "linear");
```
Could define constants?

jquick-axway committed:
Android: Updated [TIMOB-9366] code based on feedback.

- Also modified "TiUIView" to log exception thrown by "TiGradientDrawable" constructor. Particularly for invalid "type" parameters.

jquick-axway commented:
Updated PR based on review feedback.

Also modified `TiUIView` to log an exceptions thrown by `TiGradientDrawable` constructor. Particularly needed to log invalid `type` gradient string IDs.

jquick-axway committed:
Merge branch 'master' into TIMOB-9366

jquick-axway committed:
Revert "common.xml" build file change made in [TIMOB-29366]
- Was committed by mistake

jquick-axway committed:
Android: Fixed radial gradient crash caused by setting radius to zero.

jquick-axway committed:
Android: Updated code formatting for [TIMOB-9366]

jquick-axway committed:
Android: Made additional radial gradient changes for [TIMOB-9366]

- Fixed bug where if start radius is zero, a single transparent pixel can be seen in center of circle.
- Fixed bug where Android 4.2-4.4 shows glitchy graphics when start radius is greater than 0.0 and backfillStart is false.
  * This was a Google bug that had to be worked-around by adding a color offset starting from 0.0 (center of circle).
- Fixed bug where outer edge of circle would wrongly back-fill if radius was greater than 255 and if border was applied to view (HW acceleration disabled) or GPU shader is limited to "lowp" 1 byte floats.

garymathews reviewed (approved):
CR: **PASS**

👍

lokeshchdhry commented:
FR Passed.

Used the test code in the ticket & `Start Radius`, ``End Radius`, `Back Fill End` & `Back Fill Start` works as expected in various scenarios.

Studio Ver:     5.0.0.201712081732
SDK Ver:        7.2.0 local build
OS Ver:         10.13.2
Xcode Ver:      Xcode 9.2
Appc NPM:       4.2.12
Appc CLI:       7.0.2
Daemon Ver:     1.0.1
Ti CLI Ver:     5.0.14
Alloy Ver:      1.11.0
Node Ver:       8.9.1
NPM Ver:        5.5.1
Java Ver:       1.8.0_101
Devices:        ⇨ google Nexus 6P --- Android 8.0.0
                ⇨ google Nexus 5 --- Android 6.0.1
Emulator: Android 4.1, 4.2.2

Commit:
Merge branch 'master' into TIMOB-9366
